# 02 · Cleaning missing values

Keep all passenger records and make three explicit cleaning decisions.

In [1]:
from pathlib import Path
import pandas as pd

root = Path.cwd().parent if Path.cwd().name == "notebooks" else Path.cwd()
raw = pd.read_csv(root / "data" / "Titanic-Dataset.csv")

In [2]:
missing = raw.isna().sum()
pd.DataFrame({"Missing count": missing, "Missing %": (missing / len(raw) * 100).round(2)}).query("`Missing count` > 0")

,Missing count,Missing %
Age,177,19.87
Cabin,687,77.10
Embarked,2,0.22


- **Age → median:** a simple replacement that is less sensitive to extreme ages; it reduces observed variation.
- **Cabin → drop:** 77.10% is missing, so assigning cabin locations would require strong assumptions.
- **Embarked → mode:** only two values are missing; use the most frequent departure port.

In [3]:
cleaned = raw.copy()
cleaned["Age"] = cleaned["Age"].fillna(cleaned["Age"].median())
cleaned["Embarked"] = cleaned["Embarked"].fillna(cleaned["Embarked"].mode().iloc[0])
cleaned = cleaned.drop(columns="Cabin")
print("New shape:", cleaned.shape)
print("Remaining missing values:", int(cleaned.isna().sum().sum()))

New shape: (891, 11)
Remaining missing values: 0


In [4]:
cleaned.to_csv(root / "data" / "cleaned.csv", index=False)
print("Saved data/cleaned.csv for notebooks 03–05.")

Saved data/cleaned.csv for notebooks 03–05.
